cairo/ethereum/cancun/trie.cairo

```diff
@@ -134,45 +207,87 @@ from ethereum.cancun.transactions import LegacyTransaction
 //     // return trie._data.get(key, trie.default)
 // }
 
-// func common_prefix_length(a: Sequence, b: Sequence) -> felt {
-//     // Implementation:
-//     // for i in range(len(a)):
-//     // if i >= len(b) or a[i] != b[i]:
-//     // return i
-//         // if i >= len(b) or a[i] != b[i]:
-//         // return i
-//             // return i
-//     // return len(a)
-// }
+func common_prefix_length(a: Bytes, b: Bytes) -> felt {
+    alloc_locals;
+    local result;
 
-// func nibble_list_to_compact(x: Bytes, is_leaf: bool) -> Bytes {
-//     // Implementation:
-//     // compact = bytearray()
-//     // if len(x) % 2 == 0:
-//     // compact.append(16 * (2 * is_leaf))
-//     // for i in range(0, len(x), 2):
-//     // compact.append(16 * x[i] + x[i + 1])
-//     // else:
-//     // compact.append(16 * (2 * is_leaf + 1) + x[0])
-//     // for i in range(1, len(x), 2):
-//     // compact.append(16 * x[i] + x[i + 1])
-//         // compact.append(16 * (2 * is_leaf))
-//         // for i in range(0, len(x), 2):
-//         // compact.append(16 * x[i] + x[i + 1])
-//             // compact.append(16 * x[i] + x[i + 1])
-//     // else:
-//         // compact.append(16 * (2 * is_leaf + 1) + x[0])
-//         // for i in range(1, len(x), 2):
-//         // compact.append(16 * x[i] + x[i + 1])
-//             // compact.append(16 * x[i] + x[i + 1])
-//     // return Bytes(compact)
-// }
+    %{ memory[fp] = oracle(ids, reference="ethereum.cancun.trie.common_prefix_length") %}
+
+    jmp common if result != 0;
+    jmp diff;
+
+    common:
+    let index = [ap - 1];
+    with_attr error_message("common_prefix_length") {
+        assert a.value.data[index - 1] = b.value.data[index - 1];
+    }
+    tempvar index = index - 1;
+    jmp common if index != 0;
+
+    diff:
+    let result = [fp];
+    if (a.value.len == result) {
+        return result;
+    }
+
+    if (b.value.len == result) {
+        return result;
+    }
+
+    with_attr error_message("common_prefix_length") {
+        assert_not_zero(a.value.data[result] - b.value.data[result]);
+    }
+
+    return result;
+}
+
+func nibble_list_to_compact(x: Bytes, is_leaf: bool) -> Bytes {
+    alloc_locals;
+    let (local compact) = alloc();
+
+    if (x.value.len == 0) {
+        assert [compact] = 16 * (2 * is_leaf.value);
+        tempvar result = Bytes(new BytesStruct(compact, 1));
+        return result;
+    }
+
+    local remainder = nondet %{ ids.x.value.len % 2 %};
+    assert remainder * (1 - remainder) = 0;
```

Comment:
what constrains remainder to be the correct value here ?

couldn't a malicious prover invert the `remainder` value returned ?

Reply:
perhaps we should do something like:
```
local remainder
local quotient
%{memory[fp], memory[fp+1] = oracle...()%}
assert quotient * 2 + remainder == ids.x.value.len ?
```

Reply:
it doesn't change anything, the below assertion is valid for both 0 and 1

See the python implementation

```
    if len(x) % 2 == 0:  # ie even length
        compact.append(16 * (2 * is_leaf))
    else:
        compact.append(16 * ((2 * is_leaf) + 1) + x[0])
```

Reply:
```
E           starkware.cairo.lang.vm.vm_exceptions.VmException: :272:45: Error at pc=0:1777:
E           Unknown value for memory cell at address 7:2.
E               assert [compact] = 16 * x_ptr.data[i] + x_ptr.data[i + 1];
E                                                       ^***************^
E           Cairo traceback (most recent call last):
E           :115:19: (pc=0:1569)
E               let compact = nibble_list_to_compact(node.value.extension_node.value.key_segment, bool(0));
```


```
    @given(node=...)
    @settings(deadline=None, max_examples=5)
    def test_encode_internal_node_should_fail(
        self, cairo_program, cairo_run, node: Optional[InternalNode]
    ):
        with (
            patch_hint(
                cairo_program,
                'ids.remainder = ids.x.value.len % 2',
                "ids.remainder = 1",
            )
        ):
            assert sequence_equal(
                encode_internal_node(node), cairo_run("encode_internal_node", node)
            )
 ```
 
 Can't you fill the memory with whatever you want here ?